File: 字符串/2. 反转字符串 - 541/Reverse String II.cpp
/*
    23/03/28
    https://leetcode.cn/problems/reverse-string-ii/
 */

#include <iostream>
#include <string>

using namespace std;

/*
    solution: 双指针
    result: ✅
    TC: O(n)  n(k+1)/2k
    SC: O(1)
 */

string reverseStr(string s, int k) {

    for (int leftPointer = 0; leftPointer < s.size(); leftPointer += 2 * k) {
        if (leftPointer + k < s.size()) {
            reverse(s.begin() + leftPointer, s.begin() + leftPointer + k);
        } else {
            reverse(s.begin() + leftPointer, s.end());
        }
    }

    return s;
        
}

File: 字符串/2. 反转字符串 - 541/Reverse String II.py
# 23/03/28
# https://leetcode.cn/problems/reverse-string-ii/


# 双指针
# ✅

def reverseStr(s: str, k: int) -> str:

    leftPointer = 0

    while leftPointer < len(s):
        rightPointer = leftPointer + k
        s = s[:leftPointer] + s[leftPointer:rightPointer][::-1] + s[rightPointer:]
        leftPointer += 2 * k

    return s